packages/multi-test/src/app.rs

```diff
@@ -903,4 +911,172 @@ mod test {
+                    },
+                    &[],
+                )
+                .unwrap();
+
+            assert_eq!(response.data, Some("Second".to_owned().into_bytes().into()));
+        }
+
+        #[test]
+        fn single_no_data() {
+            let mut app = binary_app();
+
+            let owner = Addr::unchecked("owner");
+
+            let contract_id = app.store_code(echo::contract());
+            let contract = app
+                .instantiate_contract(contract_id, owner.clone(), &EmptyMsg {}, &[], "Echo", None)
+                .unwrap();
+
+            let response = app
+                .execute_contract(
+                    owner,
+                    contract.clone(),
+                    &echo::Message {
+                        data: Some("First".to_owned()),
+                        sub_msg: vec![make_echo_submsg(contract, None, vec![])],
+                    },
+                    &[],
+                )
+                .unwrap();
+
+            assert_eq!(response.data, Some("First".to_owned().into_bytes().into()));
+        }
+
+        #[test]
+        fn multiple_submsg() {
+            let mut app = binary_app();
+
+            let owner = Addr::unchecked("owner");
+
+            let contract_id = app.store_code(echo::contract());
+            let contract = app
+                .instantiate_contract(contract_id, owner.clone(), &EmptyMsg {}, &[], "Echo", None)
+                .unwrap();
+
+            let response = app
+                .execute_contract(
+                    owner,
+                    contract.clone(),
+                    &echo::Message {
+                        data: Some("Orig".to_owned()),
+                        sub_msg: vec![
+                            make_echo_submsg(contract.clone(), None, vec![]),
+                            make_echo_submsg(contract.clone(), "First", vec![]),
+                            make_echo_submsg(contract.clone(), "Second", vec![]),
+                            make_echo_submsg(contract, None, vec![]),
+                        ],
+                    },
+                    &[],
+                )
+                .unwrap();
+
+            assert_eq!(response.data, Some("Second".to_owned().into_bytes().into()));
+        }
+
+        #[test]
+        fn nested_submsg() {
+            let mut app = binary_app();
+
+            let owner = Addr::unchecked("owner");
+
+            let contract_id = app.store_code(echo::contract());
+            let contract = app
+                .instantiate_contract(contract_id, owner.clone(), &EmptyMsg {}, &[], "Echo", None)
+                .unwrap();
+
+            let response = app
+                .execute_contract(
+                    owner,
+                    contract.clone(),
+                    &echo::Message {
+                        data: Some("Orig".to_owned()),
+                        sub_msg: vec![make_echo_submsg(
```

Comment:
Another great test case.
I had to read it twice to ensure the final result made sense, but yes, very correct